include/circt/Support/InstanceGraph.h

```diff
@@ -51,7 +55,12 @@ class InstanceRecord
     : public llvm::ilist_node_with_parent<InstanceRecord, InstanceGraphNode> {
 public:
   /// Get the instance-like op that this is tracking.
-  HWInstanceLike getInstance() const { return instance; }
+  template <typename TTarget = InstanceOpInterface>
```

Comment:
I was pondering this functionality too.  An alternate would be an explicitly specialized implementation for InstanceOpInterface

Reply:
I also considered simply overloading the function in the per-dialect versions, but that gets a bit messy (though would eliminate the need for the op interfaces).